#Zea Maize CAGE data analysis
Raw data set:<br/>
1)Reference genome Zea_mays.AGPv3.31.dna_rm.genome.fa 
 523 scaffolds in total around 2 GB
 
2)CAGE reads for shoots and roots tissue<br/>
SRR2078289.fastq  SRR2078290.fastq for shoot<br/>
SRR2078291.fastq  SRR2078292.fastq for root<br/>

(single-end 50bp reads around 2.2GB in average)


##Mapping CAGE data back to ref genome

Use bwa alignment tools to map CAGE tags against reference genome. 

Input: ref genome, CAGE reads
Output: alignment files in bam format

##CAGEr pipeline analysis
Use CAGEr pipeline tools and tag alignment files to identify Transcription Start Sites (TSS) clusters.

Input: ref genome, alignment files (bam)
Output: clustered tags for de novo motif prediction

procedure:<br/>
1)Use ref genome fasta file to forge a BSgenome package which allows efficient genome searching for clustered tags.(forgeBSgenomeDataPkg)<br/>

2)CAGEr pipeline<br/>
(I modified the script for my data on my laptop and mason server, but unexpected error occured and we could not fix that so Taylor helped me finish this part on gremini)

##Homer motif prediction with cross-validation

Input:CAGEr clustered tags
Output: de novo predicted motifs in homer format
program: homer (findMotifsGenome.pl <pos tags> <ref Genome> Motifs.homer -len 6,8,10,12 -size -60,40)

Clustered tags(12108 in total) for four data sets are converged into one file and randomly split into 10 groups(9 training sets and 1 test set).

Use same parameter introduced above to predict motifs in these regions.Then, apply our [work flow](https://github.com/xianyao710/YRB_Capstone) to generate non-redudant motifs for each group. In theory, these motifs in 9 training sets can be observed in test.

## Tmp result
Since we did 10-fold cross validation, we should repeat our work flow 10 times. Right now, I finished the first experiment with some results.

There are 474 motifs identified by Homer from 9 training sets. Using tomtom to do comparison, and networkx package for graphical analysis, only 1 connected compoment having more than 9 nodes were observed.

![Cluster1][figure1]
[figure1]:https://github.com/xianyao710/YRB_Capstone/blob/master/logs/cluster1.png

After using MotifReduce.pl to generate consensus motifs, we get the SeqLogo for motif group 1.
![Motif1][figure2]
[figure2]:https://github.com/xianyao710/YRB_Capstone/blob/master/logs/motif1.png
  


